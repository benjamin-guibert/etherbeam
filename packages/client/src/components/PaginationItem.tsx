import React, { FC } from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import {
  faAngleDoubleLeft,
  faAngleDoubleRight,
  faAngleLeft,
  faAngleRight,
  faEllipsisH,
  IconDefinition,
} from '@fortawesome/free-solid-svg-icons'
import ListItem from './ListItem'
import Button from './Button'

type PageProp = 'first' | 'previous' | 'ellipsis' | 'next' | 'last' | number

interface PaginationItemProps {
  page: PageProp
  action?: () => void
  selected?: boolean
  disabled?: boolean
}

const PaginationItem: FC<PaginationItemProps> = ({ page, action, selected, disabled }) => {
  const getProps = (): { description: string; icon: IconDefinition } => {
    switch (page) {
      case 'first':
        return { description: 'First page', icon: faAngleDoubleLeft }
      case 'previous':
        return { description: 'Previous page', icon: faAngleLeft }
      case 'ellipsis':
        return { description: null, icon: faEllipsisH }
      case 'next':
        return { description: 'Next page', icon: faAngleRight }
      case 'last':
        return { description: 'Last page', icon: faAngleDoubleRight }
      default:
        return {
          description: `Page ${page}`,
          icon: null,
        }
    }
  }

  const { description, icon } = getProps()

  return (
    <ListItem>
      <Button link color="secondary" description={description} action={action} disabled={disabled} active={selected}>
        {typeof page === 'number' ? page : <FontAwesomeIcon icon={icon} />}
      </Button>
    </ListItem>
  )
}

export default PaginationItem
