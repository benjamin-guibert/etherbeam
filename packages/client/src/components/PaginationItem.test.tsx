import React from 'react'
import { shallow, ShallowWrapper } from 'enzyme'
import PaginationItem from './PaginationItem'

describe('<PaginationItem />', () => {
  const onClick = jest.fn()

  let component: ShallowWrapper

  describe('Default', () => {
    beforeAll(() => {
      component = shallow(<PaginationItem page={5} onClick={onClick} />)
    })

    it('should render', () => {
      const link = component.find('Link')
      expect(link.prop('title')).toBe('Page 5')
      expect(link.prop('onClick')).toBe(onClick)
      expect(link.prop('disabled')).toBeFalsy()
      expect(link.prop('active')).toBeFalsy()
      expect(link.prop('children')).toBe(5)
    })

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('First', () => {
    beforeAll(() => {
      component = shallow(<PaginationItem page="first" onClick={onClick} />)
    })

    it('should render', () => {
      const link = component.find('Link')
      expect(link.prop('title')).toBe('First page')
      expect(link.find('FontAwesomeIcon').at(0).exists()).toBeTruthy()
    })
  })

  describe('Previous', () => {
    beforeAll(() => {
      component = shallow(<PaginationItem page="previous" onClick={onClick} />)
    })

    it('should render', () => {
      const link = component.find('Link')
      expect(link.prop('title')).toBe('Previous page')
      expect(link.find('FontAwesomeIcon').at(0).exists()).toBeTruthy()
    })
  })

  describe('Ellipsis', () => {
    beforeAll(() => {
      component = shallow(<PaginationItem page="ellipsis" onClick={onClick} />)
    })

    it('should render', () => {
      const link = component.find('Link')
      expect(link.prop('title')).toBeNull()
      expect(link.find('FontAwesomeIcon').at(0).exists()).toBeTruthy()
    })
  })

  describe('Next', () => {
    beforeAll(() => {
      component = shallow(<PaginationItem page="next" onClick={onClick} />)
    })

    it('should render', () => {
      const link = component.find('Link')
      expect(link.prop('title')).toBe('Next page')
      expect(link.find('FontAwesomeIcon').at(0).exists()).toBeTruthy()
    })
  })

  describe('Last', () => {
    beforeAll(() => {
      component = shallow(<PaginationItem page="last" onClick={onClick} />)
    })

    it('should render', () => {
      const link = component.find('Link')
      expect(link.prop('title')).toBe('Last page')
      expect(link.find('FontAwesomeIcon').at(0).exists()).toBeTruthy()
    })
  })

  describe('Selected', () => {
    beforeAll(() => {
      component = shallow(<PaginationItem page={5} onClick={onClick} selected />)
    })

    it('should render', () => {
      const link = component.find('Link')
      expect(link.prop('disabled')).toBeFalsy()
      expect(link.prop('active')).toBeTruthy()
    })
  })

  describe('Disabled', () => {
    beforeAll(() => {
      component = shallow(<PaginationItem page={5} onClick={onClick} disabled />)
    })

    it('should render', () => {
      const link = component.find('Link')
      expect(link.prop('disabled')).toBeTruthy()
      expect(link.prop('active')).toBeFalsy()
    })
  })

  describe('Selected & disabled', () => {
    beforeAll(() => {
      component = shallow(<PaginationItem page={5} onClick={onClick} selected disabled />)
    })

    it('should render', () => {
      const link = component.find('Link')
      expect(link.prop('disabled')).toBeTruthy()
      expect(link.prop('active')).toBeTruthy()
    })
  })
})
