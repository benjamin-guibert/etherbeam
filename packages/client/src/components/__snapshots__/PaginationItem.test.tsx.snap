// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PaginationItem /> Default should match snapshot 1`] = `
<li>
  <a
    class="my-link"
    tabindex="0"
    title="Page 5"
  >
    5
  </a>
</li>
`;
